File: ccwap/frontend/src/api/workflows.ts
import { useQuery } from '@tanstack/react-query'
import { apiFetch, buildQuery } from './client'
import { workflowKeys } from './keys'
import type { DateRange } from '@/hooks/useDateRange'

export interface UserTypeBreakdown {
  user_type: string
  sessions: number
  total_cost: number
  total_turns: number
}

export interface UserTypeTrend {
  date: string
  user_type: string
  sessions: number
  cost: number
}

export interface AgentTreeNode {
  session_id: string
  project_display: string
  user_type: string
  total_cost: number
  children: AgentTreeNode[]
}

export interface ToolSequence {
  sequence: string[]
  count: number
  pct: number
}

export interface WorkflowResponse {
  user_types: UserTypeBreakdown[]
  user_type_trend: UserTypeTrend[]
  agent_trees: AgentTreeNode[]
  tool_sequences: ToolSequence[]
}

export function useWorkflows(dateRange: DateRange) {
  return useQuery({
    queryKey: workflowKeys.data(dateRange),
    queryFn: () => apiFetch<WorkflowResponse>(
      `/workflows${buildQuery({ from: dateRange.from, to: dateRange.to })}`
    ),
  })
}

File: ccwap/server/models/workflows.py
"""Pydantic models for workflows API."""

from typing import List

from pydantic import BaseModel


class UserTypeBreakdown(BaseModel):
    user_type: str
    sessions: int
    total_cost: float
    total_turns: int


class AgentTreeNode(BaseModel):
    session_id: str
    project_display: str
    user_type: str
    total_cost: float
    children: List['AgentTreeNode'] = []


class ToolSequence(BaseModel):
    sequence: List[str]
    count: int
    pct: float


class UserTypeTrend(BaseModel):
    date: str
    user_type: str
    sessions: int
    cost: float


class WorkflowResponse(BaseModel):
    user_types: List[UserTypeBreakdown]
    user_type_trend: List[UserTypeTrend]
    agent_trees: List[AgentTreeNode]
    tool_sequences: List[ToolSequence]
